test/functional/feature_bind_extra.py

```diff
@@ -0,0 +1,117 @@
+        # Avoid any -bind= on the command line. Force the framework to avoid
+        # adding -bind=127.0.0.1.
+        self.setup_clean_chain = True
+        self.bind_to_localhost_only = False
+        self.num_nodes = 4
+
+    def setup_network(self):
+        # Override setup_network() because we want to put the result of
+        # p2p_port() in self.extra_args[], before the nodes are started.
+        # p2p_port() is not usable in set_test_params() because PortSeed.n is
+        # not set at that time.
+
+        # Due to OS-specific network stats queries, we only run on Linux.
+        self.log.info("Checking for Linux")
+        if not sys.platform.startswith('linux'):
+            raise SkipTest("This test can only be run on Linux.")
+
+        self.have_ipv6 = test_ipv6_local()
+
+        any_ipv4 = addr_to_hex('0.0.0.0')
+        any_ipv6 = addr_to_hex('::')
+        loopback_ipv4 = addr_to_hex('127.0.0.1')
+        loopback_ipv6 = addr_to_hex('::1')
+
+        # Start custom ports after p2p and rpc ports.
+        port = PORT_MIN + 2 * PORT_RANGE
+
+        # Array of tuples [command line arguments, expected bind addresses].
+        self.expected = []
+
+        # Node0, no -bind, expected to bind on any + tor target.
+        self.expected.append(
+            [
+                [],
+                [(any_ipv4, p2p_port(0)), (loopback_ipv4, REGTEST_TOR_TARGET_PORT)]
+            ]
+        )
+        if self.have_ipv6:
+            self.expected[0][1].append((any_ipv6, p2p_port(0)))
+
+        # Node1, no -bind=...=onion, thus no extra port for Tor target.
+        self.expected.append(
+            [
+                ['-bind=127.0.0.1:{}'.format(port)],
+                [(loopback_ipv4, port)]
+            ],
+        )
+        port += 1
+
+        # Node2, no normal -bind, thus only the Tor target.
+        self.expected.append(
+            [
+                ['-bind=127.0.0.1:{}=onion'.format(port)],
+                [(loopback_ipv4, port)]
+            ],
+        )
+        port += 1
+
+        # Node3, both -bind and -bind=...=onion.
+        self.expected.append(
+            [
+                ['-bind=127.0.0.1:{}'.format(port), '-bind=127.0.0.1:{}=onion'.format(port + 1)],
+                [(loopback_ipv4, port), (loopback_ipv4, port + 1)]
+            ],
+        )
+        port += 2
+
+        # Add RPC ports to the list of expected ports to bind to for all nodes.
+        # They are not relevant for this test.
+        for i in range(len(self.expected)):
+            self.expected[i][1].append((loopback_ipv4, rpc_port(i)))
+            if self.have_ipv6:
+                self.expected[i][1].append((loopback_ipv6, rpc_port(i)))
+
+        self.extra_args = list(map(lambda e: e[0], self.expected))
+        self.setup_nodes()
+
+    def run_test(self):
+        for i in range(len(self.expected)):
+            pid = self.nodes[i].process.pid
+            assert_equal(set(get_bind_addrs(pid)), set(self.expected[i][1]))
```

Comment:
This assertion fails for me with errors like
```
AssertionError: not(
{('00000000', 11340), ('7f000001', 16340), ('00000000000000000000000000000000', 11340), ('7f000001', 18445)} ==
{('00000000', 11340), ('7f000001', 16340), ('7f000001', 18445)})
```
This fixes it, but it's not pretty and might be treating the symptoms more than the problem:
```diff
--- a/test/functional/test_framework/netutil.py
+++ b/test/functional/test_framework/netutil.py
@@ -81,7 +81,10 @@ def get_bind_addrs(pid):
     bind_addrs = []
     for conn in netstat('tcp') + netstat('tcp6'):
         if conn[3] == STATE_LISTEN and conn[4] in inodes:
-            bind_addrs.append(conn[1])
+            if conn[1][0] == '00000000000000000000000000000000':
+                bind_addrs.append(('00000000', conn[1][1]))
+            else:
+                bind_addrs.append(conn[1])
     return bind_addrs
```

Reply:
Finally I figured out what is going on!

`test_ipv6_local()` tries to connect to `::1` in order to check if the system has IPv6 support.

Some (all?) Travis machines do support IPv6, but `::1` is not configured on the loopback interface (!?). On those machines connecting to `::1` fails with `[Errno 99] Cannot assign requested address`. However it is possible to successfully bind on `::`.

`bitcoind` tries to bind on `::` and `::1` and the above makes it difficult for the test to set reasonable expectations.

Resolved by ignoring all IPv6 addresses.